cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/FastMap.scala

```diff
@@ -0,0 +1,428 @@
+package com.twitter.cassovary.util.collections
+
+import java.util
+
+import it.unimi.dsi.fastutil.ints._
+import it.unimi.dsi.fastutil.objects._
+
+import scala.collection.generic.CanBuildFrom
+import scala.collection.mutable
+import scala.{specialized => sp}
+
+/**
+ * Specialized, fast and memory efficient implementation of mutable map.
+ *
+ * Provides basic operations only. See [[FastMapWrapper]] for how to use
+ * it as a [[mutable.Map]].
+ *
+ * Use constructor provided in the companion object.
+ */
+trait FastMap[@sp(Int, Long, AnyRef) A, @sp(Int, Long, AnyRef) B] {
+
+  /**
+   * Adds key-value pair to the map.
+   */
+  def +=(key: A, value: B): this.type
+
+  /**
+   * Removes key-value pair with a given key.
+   */
+  def -=(key: A): this.type
+
+  /**
+   * Get is optimized for retrieving elements contained in the Map.
+   *
+   * If the element is not in the map, returns `0` of given type (or false, null).
+   */
+  def get(key: A): B
+
+  def clear(): Unit
+
+  def contains(key: A): Boolean
+
+  def size(): Int
+
+  /**
+   * Wraps as a [[mutable.Map]].
+   */
+  def asScala(): mutable.Map[A, B] = new FastMapWrapper[A, B](this)
+
+  /**
+   * Returns iterator over key, value pairs stored in the map.
+   *
+   * Used only inside to conform to Scala's collections.
+   */
+  private[collections] def iterator: Iterator[(A, B)]
+}
+
+/**
+ * Implements incrementing value specified by a given key.
+ */
+trait AddTo[@sp(Int, Long, AnyRef) A] {
+  def addTo(key: A, increment: Int)
+}
+
+/**
+ * Marks a [[FastMap]] with an iterator that preserves insertion order.
+ */
+trait InsertionOrderIterator
+
+/**
+ * Features creation of specialized FastMaps.
+ */
+object FastMap {
```

Comment:
make this extend FastMapFactory[FastMap] as this object really is a factory
where 
trait FastMapFactory[MapType[A,B] <: FastMap[A,B]] { def apply[@sp(...) A,@sp(...) B](): MapType }

Kind of like scala/collection/Map.scala and scala/collection/generic/MapFactory.scala